pkg/cmd/build/build.go

```diff
@@ -212,67 +212,172 @@ func translateDockerErr(err error) error {
+func (opts *BuildOptions) optimizedGlobalBuild() (string, bool, error) {
+	oktetoLog.Debugf("Applying global build optimization for image %s", opts.Tag)
+
+	globalReference := opts.Tag
+	if registry.IsDevRegistry(opts.Tag) || registry.IsExtendedOktetoRegistry(opts.Tag) {
+		globalReference = registry.ReplaceTargetRepository(opts.Tag, okteto.GlobalRegistry, okteto.DefaultGlobalNamespace)
+
 	}
-	opts.OutputMode = setOutputMode(outputMode)
+	tagWithDigest, err := registry.GetImageTagWithDigest(globalReference)
+	if err != nil {
+		return opts.Tag, false, err
+	}
+	oktetoLog.Debugf("skipping build: image %s is already built", opts.Tag)
+	opts.Tag = globalReference
+	return tagWithDigest, true, nil
+}
 
-	return opts
+// checkImageAtRegistry returns if build should be optimized because the image is already at the dev registry
+func (opts *BuildOptions) checkImageAtRegistry() (string, bool, error) {
+	oktetoLog.Debugf("Checking registry for image %s", opts.Tag)
+	tagWithDigest, err := registry.GetImageTagWithDigest(opts.Tag)
+	if err != nil {
+		return opts.Tag, false, err
+	}
+	oktetoLog.Debugf("skipping build: image %s is already built", opts.Tag)
+	return tagWithDigest, true, nil
 }
 
-// ShouldOptimizeBuild returns if optimization should be applied
-func ShouldOptimizeBuild(image string) bool {
+// SkipBuild returns if build has to be skipped and the tag with digest if found at registry
+func (opts *BuildOptions) SkipBuild(service string) (string, bool, error) {
+	if opts.NoCache {
+		return "", false, nil
+	}
+	if !registry.IsOktetoRegistry(opts.Tag) {
+		return "", false, nil
+	}
 	envGitCommit := os.Getenv(model.OktetoGitCommitEnvVar)
 	isLocalEnvGitCommit := strings.HasPrefix(envGitCommit, model.OktetoGitCommitPrefix)
-	return registry.IsOktetoRegistry(image) &&
-		envGitCommit != "" &&
-		!isLocalEnvGitCommit
+	isPipeline := envGitCommit != "" && !isLocalEnvGitCommit
+	if !isPipeline {
+		return "", false, nil
+	}
+	oktetoLog.Debug("found OKTETO_GIT_COMMIT, optimizing the build flow")
+
+	if tagWithDigest, ok, err := opts.optimizedGlobalBuild(); ok {
+		// global optimization has been applied and use global tag for deployment
+		oktetoLog.Debugf("Skipping '%s' build. Image already exists at Okteto Registry", service)
+		return tagWithDigest, true, nil
+	} else if err != nil && err != oktetoErrors.ErrNotFound {
+		// not applicated and error while checking - return err
+		return "", false, fmt.Errorf("error checking image at registry %s: %v", opts.Tag, err)
+	} else if !ok {
+		// check dev registry
+		if tagWithDigest, ok, err := opts.checkImageAtRegistry(); ok {
+			oktetoLog.Debugf("Skipping '%s' build. Image already exists at Okteto Registry", service)
+			return tagWithDigest, true, nil
+		} else if err != nil && err == oktetoErrors.ErrNotFound {
+			// image not found - have to build
+			oktetoLog.Debug("image not found, building image")
+			return "", false, nil
+		}
```

Comment:
Minor comment: as each branch has its own return, I think this reads better if you just use single `ifs` instead of `else if`

Reply:
+1, but I prefer to use switch statements on this cases instead of creating 3 different ifs